File: src/common/common/main/java/settingdust/mod_sets/data/ModSetsDisabledMods.java
package settingdust.mod_sets.data;

import com.google.gson.reflect.TypeToken;
import settingdust.mod_sets.util.LoaderAdapter;
import settingdust.mod_sets.util.SavingData;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashSet;
import java.util.Set;

public class ModSetsDisabledMods implements SavingData {
    private static final Path configPath = LoaderAdapter.get().getConfigDirectory().resolve("disabled_mods.json");
    private static Set<String> disabledMods = null;
    private static Set<String> disablingMods = null;

    public static Set<String> disabledMods() {
        return disabledMods;
    }

    public static Set<String> disablingMods() {
        return disablingMods;
    }

    @Override
    public void load() {
        try {Files.createDirectories(configPath.getParent());} catch (IOException ignored) {}
        try {
            Files.createFile(configPath);
            Files.writeString(configPath, "[]");
        } catch (IOException ignored) {}

        try {
            disabledMods = GSON.fromJson(
                Files.readString(configPath),
                (TypeToken<Set<String>>) TypeToken.getParameterized(Set.class, String.class)
            );
            if (disabledMods == null) disabledMods = new HashSet<>();
            if (disablingMods == null) disablingMods = disabledMods;
        } catch (IOException e) {
            throw new RuntimeException("Fail to load disabled mods", e);
        }
        save();
    }

    @Override
    public void save() {
        try {
            GSON.toJson(disabledMods, Files.newBufferedWriter(configPath));
        } catch (IOException e) {
            throw new RuntimeException("Fail to save disabled mods", e);
        }
    }
}

File: src/common/main/kotlin/settingdust/mod_sets/data/ModSetsDisabledMods.kt
package settingdust.mod_sets.data

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.json.decodeFromStream
import kotlinx.serialization.json.encodeToStream
import settingdust.mod_sets.ModSets
import settingdust.mod_sets.util.LoaderAdapter
import kotlin.io.path.createDirectories
import kotlin.io.path.createFile
import kotlin.io.path.div
import kotlin.io.path.inputStream
import kotlin.io.path.outputStream
import kotlin.io.path.writeText

class ModSetsDisabledMods : SavingData {
    companion object {
        var disabledMods = mutableSetOf<String>()
            private set
        lateinit var disabledModsInitial: Set<String>
            private set
        private val configPath = LoaderAdapter.configPath / "disabled_mods.json"

        private fun isDisabledModsInitialInitialized() = ::disabledModsInitial.isInitialized
    }

    @OptIn(ExperimentalSerializationApi::class)
    override fun reload() {
        runCatching { LoaderAdapter.configPath.createDirectories() }
        runCatching {
            configPath.createFile()
            configPath.writeText("[]")
        }
        disabledMods = ModSets.configJson.decodeFromStream(configPath.inputStream())
        if (!isDisabledModsInitialInitialized()) disabledModsInitial = disabledMods
        save()
    }

    @OptIn(ExperimentalSerializationApi::class)
    override fun save() {
        ModSets.configJson.encodeToStream(disabledMods, configPath.outputStream())
    }
}
